[Skia] Variation selector should take precedence over font-variant-emoji

carlosgcampos opened the pull request:
#### be26a0e3aa8b5ab63a9d4a3d96372d8307382ac2
<pre>
[Skia] Variation selector should take precedence over font-variant-emoji
<a href="https://bugs.webkit.org/show_bug.cgi?id=295097">https://bugs.webkit.org/show_bug.cgi?id=295097</a>

Reviewed by Nikolas Zimmermann.

According to the spec: &quot;Even when font-variant-emoji is used, the
presence of Variation Selector 15 (U+FE0E) or Variation Selector 16
(U+FE0F) in the contents of an element override the rendering specified
in font-variant-emoji.&quot;

Canonical link: <a href="https://commits.webkit.org/297109@main">https://commits.webkit.org/297109@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aee8e703129d974ad5e33ae17e5915e57ab65637

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110574 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30233 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20666 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116600 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60841 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112537 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30912 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38821 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84086 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113522 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24682 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99566 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64527 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24045 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17706 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60395 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94067 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17765 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119390 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37614 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27931 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93053 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37988 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95837 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92875 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37885 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15626 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33588 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17838 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37509 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42980 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37172 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40511 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38880 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/FontCascadeSkia.cpp:
```diff
@@ -128,13 +128,46 @@ RefPtr<const Font> FontCascade::fontForCombiningCharacterSequence(StringView str
     bool isOnlySingleCodePoint = codePointsIterator == codePoints.end();
 
+    bool shouldForceEmojiFont = false;
+    ResolvedEmojiPolicy emojiPolicy = [&] {
```
Looks a bit awkward that the lambda fills in the bool value. Structured binding could encapsulate this better:
```auto [emojiPolicy, shouldForceEmojiFont] = [&] {```. I was originally hoping to get rid of the capture in the lambda, but it uses quite a few other variables from the outer function, so I take back the suggestion ;-)

nikolaszimmermann reviewed (approved):
LGTM, r=me.

carlosgcampos committed:
[Skia] Variation selector should take precedence over font-variant-emoji
https://bugs.webkit.org/show_bug.cgi?id=295097

Reviewed by Nikolas Zimmermann.

According to the spec: "Even when font-variant-emoji is used, the
presence of Variation Selector 15 (U+FE0E) or Variation Selector 16
(U+FE0F) in the contents of an element override the rendering specified
in font-variant-emoji."

Canonical link: https://commits.webkit.org/297109@main
